Made self and parent their own grammar nodes

wu-s-john opened the pull request:
This is a PR that helps solves the issue, [PHP, not recognizing const variables as constant strings](https://github.com/returntocorp/semgrep/issues/3921)

- In php, self and parent are used to get particular information of a class. Self is used to help get static information of a class and parent is used to get information of the supertype of an object. It would be nice to exhibit these behaviors for semgrep. This information is useful to constant propogation. To make this more apparent, it would be helpful to have these nodes represented in the php CST. These fields represented in CST_php.class_scope_resolution_ident.
- Self and Parent are now special identifiers in the php_ast
- Essentially, self seems to be only used as identifiers in expressions and as type annotations in php. This change helps accomodate for that.

The corresponding semgrep changes are made here. https://github.com/returntocorp/semgrep/pull/5022

### Security

- [x] Change has no security implications (otherwise, ping the security team)

wu-s-john committed:
Made self and parent their own grammar nodes
- In php, self and parent are used to get particular information of a class. Self is used to help get static information of a class and parent is used to get information of the supertype of an object. It would be nice to exhibit these behaviors for semgrep. This information is useful to constant propogation. To make this more apparent, it would be helpful to have these nodes represented in the php CST. These fields represented in CST_php.class_scope_resolution_ident.
- Self and Parent are now special identifiers in the php_ast
- Essentially, self seems to be only used as identifiers in expressions and as type annotations in php. This change helps accomodate for that.

IagoAbal commented on lang_php/analyze/ast_php.ml:
```diff
@@ -233,6 +233,10 @@ and special =
   | This
   (* take many different forms in PHP, eval(), call_user_func, ${}, etc. *)
   | Eval
+  (* often transformed in Var "$self" in the analysis *)
```
Do you understand this comment or are you just copying what is being said about `This`?

IagoAbal replied:
Could you also update a comment in the top of the file that says?
>  *  - some special keywords, for instance 'self', are transformed in
> *    "__special__self". See special() below.
The comment is still relevant but we should use a different example than `self`.

IagoAbal commented on lang_php/parsing/cst_php.ml:
```diff
@@ -123,14 +123,19 @@ and qualified_ident_element =
+
 type name =
   | XName of qualified_ident
   (* Could also transform at parsing time all occurences of self:: and
```
You would need to do something about this comment.

IagoAbal commented:
I don't know much about PHP (maybe @nmote know some PHP?), but overall it looks good to me! Could you take a look at my comments? Thanks for contributing!

IagoAbal commented on lang_php/parsing/cst_php.ml:
```diff
@@ -142,6 +147,7 @@ type name =
 type hint_type =
   | Hint of name (* only self/parent, no static *) *
             type_args option
+  | HintScopeResolution of class_scope_resolution_ident * type_args option
```
Do we really need this? Can't we just use `Hint` directly? Anyways, that's what we do later in `Ast_php_build`.

IagoAbal replied:
But don't take it as a "must-do" comment... I just think it's worth considering.

wu-s-john replied:
@IagoAbal So, when `self` is used as a type parameter, are you suggesting that it gets represented as the following in the CST:

```
XName [QI (Name ("self", tok))]
```

aryx replied:
yes I don't think you need to modify the CST. The CST already has some special nodes for Self and Parent (in name). The main problem was that ast_php.ml didn't have those nodes anymore, and so the information was lost later when converting to the generic AST. Adding Self/Parent to ast_php.ml should be enough.

aryx reviewed (changes requested):
Also it's usually good to provide a Test plan: section in a PR.
How did you test your change? What is the impact on parsing statistics? 
You can see semgrep/parsing-stats/ and run ./run-lang php before and after your PR
to see how things evolved.

wu-s-john committed:
Addressing PR comments
- Reverted changes on php lexer, parser and cst code
- For expressions, we easily transform cst identifiers of "self" and "parent" to special identifier nodes

wu-s-john commented:
@aryx 

Here is the test plan and the parsing stats affected by my PR. 

In the end, the changes to the parsing code was very slight. You can see the changes below where the expression `self` is translated to `IdSpecial((Self, ())`

## Test Plan

```
$ cat  semgrep-core/tests/php/constant_propogation.php
<?php

class Example
{
    const CMD = 'This is an constant.';

    public function someMethod()
    {
        // ERROR:
        exec(self::CMD);
    }
}
```


```
$ pfff -dump_php semgrep-core/tests/php/constant_propogation.php #cst_php
[(ClassDef
    { c_attrs = None; c_type = (ClassRegular ());
      c_name = (Name ("Example", ())); c_tparams = None; c_extends = None;
      c_implements = None; c_enum_type = None;
      c_body =
      ((),
       [(ClassConstants ([], (), None,
           [(Left
               ((Name ("CMD", ())),
                ((), (Sc (C (String ("This is an constant.", ())))))))
             ],
           ()));
         (Method
            { f_attrs = None; f_tok = (); f_type = MethodRegular;
              f_modifiers = [(Public, ())]; f_ref = None;
              f_name = (Name ("someMethod", ())); f_tparams = None;
              f_params = ((), [], ()); f_return_type = None;
              f_body =
              ((),
               [(ExprStmt (
                   (Call ((Id (XName [(QI (Name ("exec", ())))])),
                      ((),
                       [(Left
                           (Arg
                              (ClassGet ((Id (Self ())), (),
                                 (Id (XName [(QI (Name ("CMD", ())))]))))))
                         ],
                       ())
                      )),
                   ()))
                 ],
               ())
              })
         ],
       ())
      });
  (FinalDef ())]
```

```
$ semgrep-core -dump_pfff_ast -lang php semgrep-core/tests/php/constant_propogation.php #ast_php
Pr(
  [DefStmt(
     ({
       name=EN(
              Id(("Example", ()),
                {id_info_id=5; id_hidden=false; id_resolved=Ref(None); 
                 id_type=Ref(None); id_svalue=Ref(None); }));
       attrs=[]; tparams=[]; },
      ClassDef(
        {ckind=(Class, ()); cextends=[]; cimplements=[]; cmixins=[]; 
         cparams=[]; 
         cbody=[F(
                  DefStmt(
                    ({
                      name=EN(
                             Id(("CMD", ()),
                               {id_info_id=1; id_hidden=false; 
                                id_resolved=Ref(None); id_type=Ref(None); 
                                id_svalue=Ref(None); }));
                      attrs=[KeywordAttr((Const, ()))]; tparams=[]; },
                     VarDef(
                       {vinit=Some(L(String(("This is an constant.", ())))); 
                        vtype=None; }))));
                F(
                  DefStmt(
                    ({
                      name=EN(
                             Id(("someMethod", ()),
                               {id_info_id=4; id_hidden=false; 
                                id_resolved=Ref(None); id_type=Ref(None); 
                                id_svalue=Ref(None); }));
                      attrs=[KeywordAttr((Public, ()))]; tparams=[]; },
                     FuncDef(
                       {fkind=(Method, ()); fparams=[]; frettype=None; 
                        fbody=FBStmt(
                                Block(
                                  [ExprStmt(
                                     Call(
                                       N(
                                         Id(("exec", ()),
                                           {id_info_id=2; id_hidden=false; 
                                            id_resolved=Ref(None); 
                                            id_type=Ref(None); 
                                            id_svalue=Ref(None); })),
                                       [Arg(
                                          DotAccess(IdSpecial((Self, ())),
                                            (),
                                            FN(
                                              Id(("CMD", ()),
                                                {id_info_id=3; 
                                                 id_hidden=false; 
                                                 id_resolved=Ref(None); 
                                                 id_type=Ref(None); 
                                                 id_svalue=Ref(None); }))))]),
                                     ())]));
                        }))))];
         })))])
```

## Stats
stats before PR change:
```
lang/php/stats.json:3
   "global": {
     "name": "*",
     "parsing_rate": 0.8984308332917063,
     "line_count": 5305045,
     "error_line_count": 538829,
     "file_count": 35023,
     "error_file_count": 1498
   },
```

stats after PR change:
```
lang/php/stats.json:3
   "global": {
     "name": "*",
     "parsing_rate": 0.8984308332917063,
     "line_count": 5305045,
     "error_line_count": 538829,
     "file_count": 35023,
     "error_file_count": 1498
   },
```

Overall, doesn't look like anything change

wu-s-john commented:
@IagoAbal @aryx 

Thanks for the suggestions. I incorporated your feedback. Please take a look again when you have time.

aryx commented:
not sure what's going on with circleci not running

wu-s-john commented:
@aryx 

Thanks for merging the PR.
